### PR TITLE
Feat/go add query to all listings

### PR DESCRIPTION
**Pull request by OlfaBensoussia**
## Description
We made this PR to be able to support adding a filtering query while listing assets and contract definitions.
This resolves #30 and suggests a refactored way to validate query inputs.

## Approach
- Updated the endpoints and the requests methods
-  Added a new package **apivalidator** under **internal** to avoid copy-pasting the same object and function definition in multiple files.
- Renamed `ListPoliciesInput` to `QueryInput` because we will be referring to the same object in all of the **POST** requests [Check [this link ](https://app.swaggerhub.com/apis/eclipse-edc-bot/management-api/0.0.1-SNAPSHOT#/QuerySpecDto)for reference]
- Renamed `validateQueryPoliciesInput` to `ValidateQueryInput`
- Fixed what seemed to be some typos in the returned error messages

### Open Questions and Pre-Merge TODOs
- I was thinking that we could move `ApiErrorDetail` to the **apivalidator** package too and use it for the other entities as well (assets & contract definitions).

### CODE REVIEW

**Comment by OlfaBensoussia** [comment]
@NabilHouidi @AhmedGrati Tagging you guys here because I know that Youssef will be OOO until Thursday. 
 I would really appreciate it if you could take some time to review those changes and provide any feedback or suggestions for improvement.
I understand that you may be busy, so please feel free to take your time with this.

**Review by AhmedGrati — COMMENTED** [review]
It's an LGTM for me. Great job @OlfaBensoussia 👌
Just one small comment, can we add some unit tests to the `apivalidator`?
As for the API error, @NabilHouidi opened a PR to unify errors, so we'll get rid of the API errors.

**Comment by AhmedGrati** [comment]
@OlfaBensoussia Can you please also update the `examples/assetsCrud/main.go` file with the introduced changes?

**Comment by OlfaBensoussia** [comment]
@AhmedGrati I added everything and updated a few things all along the way.
The git history is not the cleanest but I will make sure to take care of that once I have your approval.

File: go/common/apivalidator/validate_query.go
Comment (AhmedGrati) [inline go/common/apivalidator/validate_query.go:14]: @OlfaBensoussia we added a new way of handling errors and I would suggest adapting this to the existing error handling method can be better for consistency reasons.

File: go/examples/assetsCrud/main.go
Comment (AhmedGrati) [inline go/examples/assetsCrud/main.go:184]: small nit: printing the error here can be useful.

File: go/examples/contractDefinitionsCrud/main.go
Comment (AhmedGrati) [inline go/examples/contractDefinitionsCrud/main.go:111]: Same as above.

File: go/examples/policiesCrud/main.go
Comment (AhmedGrati) [inline go/examples/policiesCrud/main.go:89]: same as above.

File: go/service/assets/api_op_ListAssets.go
Comment (AhmedGrati) [inline go/service/assets/api_op_ListAssets.go:24]: Just a question here: 
why do we take only one `queryInput`? If the connector API accepts only one query Input and not an array, I suggest updating the `ListAssets` parameter to a `*apivalidator.QueryInput` rather than an array of `apivalidator.QueryInput`. wdyt?

File: go/service/assets/api_op_ListAssets.go
Comment (AhmedGrati) [inline go/service/assets/api_op_ListAssets.go:38]: Same as above, I highly recommend updating the error handling method to the new one. You can see an example in line 42.

File: go/service/contractdefinition/api_op_ListContractDefinitions.go
Comment (AhmedGrati) [inline go/service/contractdefinition/api_op_ListContractDefinitions.go:33]: Same as above.

File: go/service/contractdefinition/api_op_ListContractDefinitions.go
Comment (AhmedGrati) [inline go/service/contractdefinition/api_op_ListContractDefinitions.go:44]: Also here I recommend using the new error-handling method. Curious to hear your thoughts about it.

File: go/service/policies/api_op_ListPolicies.go
Comment (AhmedGrati) [inline go/service/policies/api_op_ListPolicies.go:43]: Same as above.

**Review by AhmedGrati — COMMENTED** [review]
Good job @OlfaBensoussia 👌, just some small nits and comments about the error handling.
I would recommend also adding some test cases in the unit tests by passing a `QueryInput` in `List` methods of assets, policies, etc...
I invite @NabilHouidi @Nazdroth also to take a look at the PR.

File: go/service/assets/api_op_ListAssets.go
Comment (OlfaBensoussia) [reply go/service/assets/api_op_ListAssets.go:24]: That line of code checks if we have a `queryInput` in the first place because it's optional. As you mentioned, the connector API expects only one query max.
The only reason why I went with this approach is to be able to call `ListAssets()` this way rather than `ListAssets(nil)` in case we don't have a query. [You can see pictures above of how I coded it previously]
343574294_583671690412259_5966455409479978209_n

344762782_6066875723439747_4831448852812178306_n

The approach is very biased because it follows the way we do it in Javascript 😅 But I'm open to change it too.

Comment (AhmedGrati) [reply go/service/assets/api_op_ListAssets.go:24]: Yes makes more sense now, I see that your approach is better now. Thank you 🙏.

**Comment by OlfaBensoussia** [comment]
@AhmedGrati thank you for your reviews again 💯 I addressed all of the comments in my latest commit.

File: go/common/apivalidator/validate_query_test.go
Comment (AhmedGrati) [inline go/common/apivalidator/validate_query_test.go:23]: Small nit:
```suggestion
			expectedErr: sdkErrors.FailedTo(internal.ACTION_INPUT_VALIDATE),
```

**Review by AhmedGrati — APPROVED** [review]
It's an LGTM for me, just a small nit.

File: go/service/assets/api_op_ListAssets.go
Comment (OlfaBensoussia) [reply go/service/assets/api_op_ListAssets.go]: Taken into consideration

**Pull request merged**
